core/src/main/scala/zio/config/ConfigSource.scala

```diff
@@ -1,166 +1,108 @@
 package zio.config
 
-import zio.{ system, IO, ZIO }
-import zio.system.System
 import java.{ util => ju }
+import zio.UIO
 
-final case class ConfigValue[A](value: ::[A])
+import scala.collection.JavaConverters._
+import scala.collection.immutable.Nil
+import zio.config.PropertyTree.unflatten
 
 final case class ConfigSource[K, V](
-  getConfigValue: Vector[K] => IO[ReadErrorsVector[K, V], ConfigValue[V]],
+  getConfigValue: Vector[K] => PropertyTree[K, V],
   sourceDescription: List[String]
 ) { self =>
   final def orElse(that: => ConfigSource[K, V]): ConfigSource[K, V] =
     ConfigSource(
-      k => getConfigValue(k).orElse(that.getConfigValue(k)),
-      self.sourceDescription ++ that.sourceDescription
+      k => getConfigValue(k).getOrElse(that.getConfigValue(k)),
+      if (self.sourceDescription == that.sourceDescription) self.sourceDescription
+      else self.sourceDescription ++ that.sourceDescription
     )
 
-  final def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] = self orElse that
+  final def <>(that: => ConfigSource[K, V]): ConfigSource[K, V] =
+    self orElse that
 }
 
 object ConfigSource {
   val SystemEnvironment = "system environment"
   val SystemProperties  = "system properties"
-  val JavaProperties    = "java properties"
-  val ConstantMap       = "constant <map>"
-  val EmptySource       = "<empty>"
 
   def empty[K, V]: ConfigSource[K, V] =
-    ConfigSource((k: Vector[K]) => IO.fail(singleton(ReadError.MissingValue(k))), EmptySource :: Nil)
+    ConfigSource(_ => PropertyTree.empty, Nil)
 
-  def fromEnv(valueSeparator: Option[String] = None): ConfigSource[String, String] =
-    ConfigSource(
-      (path: Vector[String]) => {
-        val key = path.mkString("_")
-        system
-          .env(key)
-          .bimap(
-            ReadError.FatalError(path, _),
-            opt =>
-              opt.map(r => {
-                val consOfValues: ::[String] =
-                  valueSeparator
-                    .flatMap(
-                      sep =>
-                        r.split(sep).toList.map(_.trim) match {
-                          case h :: tail => Some(::(h, tail))
-                          case Nil       => None
-                        }
-                    )
-                    .getOrElse(::(r, Nil))
-
-                ConfigValue(consOfValues)
-              })
-          )
-          .flatMap(IO.fromOption(_))
-          .mapError[ReadErrorsVector[String, String]](_ => singleton(ReadError.MissingValue(path)))
-          .provideLayer(System.live)
-      },
-      SystemEnvironment :: Nil
-    )
+  def fromSystemEnv(valueSeparator: Option[Char] = None): UIO[ConfigSource[String, String]] =
```

Comment:
It will lead to a nicer API if we have two versions:

```scala
def fromSystemEnv: ...
def fromSystemEnv(valueSeparator: Char)
```

This way, you don't have to wrap in `Option`, and don't have to use empty parens for the "none" case.

Any method that takes a single parameter with default argument can be factored like this.